### PR TITLE
Marvin3

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR prepares marvin and tests for python3. it was part of #4479, until nose2 was decided to be abandoned from that PR.

re-PR of #4543 and #3730  to enable cooperation

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by GabrielBrascher** [comment]
@blueorangutan package

File: systemvm/test/runtests.sh [hunk 1]
```diff
@@ -22,17 +22,19 @@ export PYTHONPATH="../debian/opt/cloud/bin/"
...
     exit 1
 fi
 
 echo "Running pylint to check systemvm/python code for errors"
-pylint --disable=R,C,W *.py
-pylint --disable=R,C,W `find ../debian -name \*.py`
+python --version
+pyenv versions
+pylint3 --disable=R,C,W *.py
+pylint3 --disable=R,C,W `find ../debian -name \*.py`
```
Comment (DaanHoogland) [inline systemvm/test/runtests.sh:37]:
```suggestion
pylint --disable=R,C,W *.py
pylint --disable=R,C,W `find ../debian -name \*.py`
```

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
reference marvin on python2 component test results:
```
# cat componenttest-summary.txt 
Test completed. 114 look ok, 57 have errors
Only failed tests results shown below:
```

**Comment by DaanHoogland** [comment]
travis green \o/ cc @GabrielBrascher @rhtyd

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by GabrielBrascher** [comment]
Manual tests are looking fine. Tested on Advanced network with Security groups.
Management node and KVM nodes are running in Ubuntu 18.04.

1. Build from source & upgrade zone with the respective packages
2. re-spawn VR & monitor cloud.log for errors
3. launch VM
4. check KVM agent log asserting no errors
5. assert that iptables (ipv4 and ipv6) rules have been applied accordingly to the security group

File: packaging/centos7/cloud.spec [hunk 2]
```diff
@@ -418,6 +424,8 @@ fi
 # Install mysql-connector-python
 pip3 install %{_datadir}/%{name}-management/setup/wheel/six-1.15.0-py2.py3-none-any.whl %{_datadir}/%{name}-management/setup/wheel/setuptools-47.3.1-py3-none-any.whl %{_datadir}/%{name}-management/setup/wheel/protobuf-3.12.2-cp36-cp36m-manylinux1_x86_64.whl %{_datadir}/%{name}-management/setup/wheel/mysql_connector_python-8.0.20-cp36-cp36m-manylinux1_x86_64.whl
 
+pip3 install urllib3
```
Comment (rhtyd) [inline packaging/centos7/cloud.spec:427]: @DaanHoogland why do we need it for cloudstack-management pkg?

**Comment by rhtyd** [comment]
@DaanHoogland pl also check/add support for python3 (urlib3 etc) for cloudstack-marvin pkg for Debian and CentOS8 - otherwise largely LGTM; I'll do another round of review. Pl also run test matrix smoketests.

File: packaging/centos7/cloud.spec (hunk 2 above)
Comment (DaanHoogland) [reply packaging/centos7/cloud.spec:427]: I'll double check, I think it's an indirect dependency, Also it is not packaged which is mayby not what we want , I would mean internet has to be available during install. I'll look into this.

**Comment by GabrielBrascher** [comment]
It is worthy to mention that my tests have been executed only on _Advanced network with Security groups_; which means that I tested the simplest VR scenario where most of the work is done on the KVM side.

We need to test other services from VR in order to verify if there are missing changes on VR scripts.
Any help with that is very welcome :-)

**Comment by DaanHoogland** [comment]
too many regressions, I'm going to call those environmental but will need to investigate :(

**Comment by DaanHoogland** [comment]
most errors explained: str vs byte[] in python3 vr scripts. need cleaning up.

**Comment by DaanHoogland** [comment]
up to 50 failing suites now, :facepalm: not sure how to address this anymore.
I think these are all due to the router scripts, but diagnostics is very bad, so I can't be sure.

File: test/integration/smoke/test_vm_life_cycle.py [hunk 3]
```diff
@@ -1443,11 +1433,13 @@ def setUp(self):
         if len(self.hosts) < 2:
             self.skipTest("Requires at least two hosts for performing migration related tests")
 
+        for host in self.hosts:
+            print(f"=== setup() === Host.OS = {host.details['Host.OS']}")
+            if host.details['Host.OS'] in ['CentOS']:
+                self.skipTest("live migration is not stabily supported on CentOS")
+
```
Comment (shwstppr) [inline test/integration/smoke/test_vm_life_cycle.py:1440]: @DaanHoogland can we have this in a separate PR to get this merged earlier to fix failing migration tests on Trillian https://github.com/apache/cloudstack/issues/4757?
Assuming Marvin3 work might need some more time.
cc @rhtyd  @weizhouapache

Comment (DaanHoogland) [reply test/integration/smoke/test_vm_life_cycle.py:1440]: will do

Comment (DaanHoogland) [reply test/integration/smoke/test_vm_life_cycle.py:1440]: see #4801

File: debian/cloudstack-marvin.postinst [hunk 4]
```diff
@@ -19,5 +19,6 @@
 
 set -e
 
-pip install --upgrade http://cdn.mysql.com/Downloads/Connector-Python/mysql-connector-python-2.0.4.zip#md5=3df394d89300db95163f17c843ef49df
-pip install --upgrade /usr/share/cloudstack-marvin/Marvin-*.tar.gz
+python3 -m pip install --upgrade pip
+python3 -m pip install --upgrade http://cdn.mysql.com/Downloads/Connector-Python/mysql-connector-python-2.0.4.zip#md5=3df394d89300db95163f17c843ef49df
```
Comment (weizhouapache) [inline debian/cloudstack-marvin.postinst:23]: @GabrielBrascher line 23 is not needed, as pip3 install latest mysql-connector-python 8.0.15
```
root@mgt01:~# pip3 list |grep mysql-connector-python
mysql-connector-python 8.0.15
```

Comment (GabrielBrascher) [reply debian/cloudstack-marvin.postinst]: Thanks for the hint, @weizhouapache. The code has been updated.

**Comment by DaanHoogland** [comment]
@GabrielBrascher @rhtyd , the travis mix of python2 and python3 does not work as we desire. I disabled the systemvm part in travis. We'll need to reenable it once we succeeded in porting the full VR - and SVM functionality.
I suggest we merge this and continue from there.

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
ping @GabrielBrascher @weizhouapache @rhtyd @shwstppr can we merge?

**Comment by pdion891** [comment]
+1

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
+1 but we should run a test round against ubuntu and CentOS8 as well
@blueorangutan help

**Comment by rhtyd** [comment]
Nevermind - I'll take my comment back, as Trilian uses CentOS7 to run tests so it wouldn't cause any change. Let's merge based on test matrix and Travis! Fantastic effort @DaanHoogland

File: pom.xml [hunk 5]
```diff
@@ -974,6 +974,7 @@
                             <exclude>**/*.md</exclude>
                             <exclude>.java-version</exclude>
                             <exclude>.python-version</exclude>
+                            <exclude>systemvm/.pythen-version</exclude>
```
Comment (resmo) [inline pom.xml:977]: @DaanHoogland typo?

Comment (DaanHoogland) [reply pom.xml:977]: yes,m definitely,  but also this file is not in use anymore.

**Pull request merged**